### PR TITLE
Add ability to create sockets early

### PR DESCRIPTION
**Pull request by saghul**
As per the LEP: https://github.com/libuv/leps/pull/5

This is still WIP, I's just like a bit of early feedback.

ToDo:
* [x] Discuss / bikeshed on the naming
* [x] Windows
* [x] UDP
* [x] More tests
* [x] Add "versionadded" to the docs

R=@bnoordhuis

### CODE REVIEW

**Comment by bnoordhuis** [comment]
I'd probably make the `domain` argument to `uv_tcp_init_ex()` a general purpose flags argument with the domain in the low bits.

There are only about 40-something address families and every platform I know of numbers them incrementally starting from zero, so you'd only need 6 bits to cover them all.  Make it 8 or 10 bits to play it safe.

(Although that's wasting a lot of bits... `AF_UNSPEC` is always zero and `AF_INET` and `AF_INET6` always < 32, AFAIK.  But that aside.)

**Comment by saghul** [comment]
Can you elaborate a bit on why? Extensibility? FWIW, this is the first time that we modify some uv_*_init in years (if memory serves right), since they tend to basically do nothing, so I'm not really concerned about that.

**Comment by bnoordhuis** [comment]
Extensibility, yes.  Always leave an escape hatch if at all possible.

**Comment by saghul** [comment]
Let's assume there is some UV_NO_CLOEXEC flag which prevents the sockets from being created with O_CLOEXEC set. This API would then be used as follows:

~~~~
uv_tcp_init_ex(loop, handle, AF_INET|UV_NO_CLOEXEC);
~~~~

Is that right? Looks a bit weird to me :-S I can add a dummy flags thing if you really feel strong about it:

~~~~
int uv_tcp_init_ex(uv_loop_t*, uv_tcp_t* handle, int domain, int flags);
uv_tcp_init_ex(loop, handle, AF_INET, UV_NO_CLOEXEC|UV_DO_AWESOME_STUFF);
~~~~

WDYT?

**Comment by bnoordhuis** [comment]
The three arg version looks alright to me an is a bit easier to remember; you can't accidentally swap the domain and the flags arguments.  You can alias AF_INET as UV_AF_INET if using plain AF_INET bothers you.

**Comment by saghul** [comment]
@bnoordhuis is https://github.com/saghul/libuv/commit/9e02f498c7c4620b2eeb1f1567ccf9c5d9c09bc9 what you were suggesting?

**Comment by saghul** [comment]
@bnoordhuis I'm chasing some failures on Windows, but can you PTAL the unix side?

File: test/test-tcp-create-socket-early.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  ASSERT(sockname.sin_family == AF_INET);
+
+  r = uv_tcp_bind(&client, (const struct sockaddr*) &addr, 0);
+  ASSERT(r == 0);
+
+  namelen = sizeof sockname;
+  r = uv_tcp_getsockname(&client, (struct sockaddr*) &sockname, &namelen);
+  ASSERT(r == 0);
+  ASSERT(memcmp(&(addr.sin_addr),
+                &(sockname.sin_addr),
```
Comment (bnoordhuis) [inline test/test-tcp-create-socket-early.c]: Femto-nit: the parens aren't necessary here.

File: test/test-tcp-create-socket-early.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                sizeof(struct in_addr)) == 0);
```
Comment (bnoordhuis) [inline test/test-tcp-create-socket-early.c]: `sizeof(addr.sin_addr)`?

File: test/test-tcp-open.c [hunk 3]
```diff
@@ -180,3 +191,30 @@ TEST_IMPL(tcp_open) {
...
+
+  startup();
+  sock1 = create_tcp_socket();
+  sock2 = create_tcp_socket();
+
+  r = uv_tcp_init(uv_default_loop(), &client);
+  ASSERT(r == 0);
+
+  r = uv_tcp_open(&client, sock1);
+  ASSERT(r == 0);
```
Comment (bnoordhuis) [inline test/test-tcp-open.c]: Suggestion: add another test asserting that calling `uv_tcp_open(&client, sock1)` again fails with UV_EBUSY.

File: test/test-udp-create-socket-early.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+  r = uv_udp_bind(&client, (const struct sockaddr*) &addr, 0);
+  ASSERT(r == 0);
+
+  namelen = sizeof sockname;
+  r = uv_udp_getsockname(&client, (struct sockaddr*) &sockname, &namelen);
+  ASSERT(r == 0);
+  ASSERT(memcmp(&(addr.sin_addr),
+                &(sockname.sin_addr),
+                sizeof(struct in_addr)) == 0);
```
Comment (bnoordhuis) [inline test/test-udp-create-socket-early.c]: `sizeof(addr.sin_addr)`?

File: test/test-udp-open.c [hunk 5]
```diff
@@ -164,3 +175,30 @@ TEST_IMPL(udp_open) {
...
+
+  startup();
+  sock1 = create_udp_socket();
+  sock2 = create_udp_socket();
+
+  r = uv_udp_init(uv_default_loop(), &client);
+  ASSERT(r == 0);
+
+  r = uv_udp_open(&client, sock1);
+  ASSERT(r == 0);
```
Comment (bnoordhuis) [inline test/test-udp-open.c]: Suggestion: add another test asserting that calling uv_tcp_open(&client, sock1) again fails with UV_EBUSY.

**Comment by bnoordhuis** [comment]
UNIX side LGTM minus some nits.  Windows side looks okay to me too.

File: test/test-tcp-open.c (hunk 3 above)
Comment (saghul) [inline test/test-tcp-open.c]: There is another thest for that :-)

**Comment by saghul** [comment]
@bnoordhuis Thanks for the review! CI run here: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/109/

**Comment by saghul** [comment]
Ok, addressed everything and the build is "green" (surprisingly we've been keeping it that way on Windows for a while \o/).

If nobody has any further comments I'll land this and the LEP later tonight.

Thanks again for the review Ben!

**Comment by saghul** [comment]
Landed!

**Comment by vavrusa** [comment]
Just a followup, can you do an API bump in the `master/v1.x` so I can #ifdef for the new version?

**Comment by saghul** [comment]
@vavrusa we bump the version when making a release (the tool does it), so you'll have to wait until we make a 1.7.0 release, sorry.

**Comment by vavrusa** [comment]
Okay, a small request then - could you make a macro in `uv-version.h` (basically `UV_VERSION` from `version.c`) so I could check for version instead of `MAJOR > 1 || MAJOR == 1 && (MINOR > 6 && ...` in preprocessor? I like both decimal `171` or hex `0x00010700` ways.

Example:
```
#if UV_VERSION >= 0x00107000
#define ENABLE_REUSEPORT
#endif
```

Decimal is neater, but it requires you must use 1 digit per minor release at most.

**Comment by saghul** [comment]
@vavrusa sounds ok to me, here is a PR: https://github.com/libuv/libuv/pull/412

**Comment by saghul** [comment]
@vavrusa you can now use UV_VERSION_HEX.

**Comment by brimworks** [comment]
BTW, the change to src/win/pipe.c in this change:

https://github.com/libuv/libuv/commit/fb5df542ace7d6735a9e8e09d7d8bb6b8abdc459

Is causing this problem:

https://github.com/luvit/luv/issues/177

Removing this code from src/win/pipe.c:uv_set_pipe_handle() fixes the problem:

  if (handle->handle != INVALID_HANDLE_VALUE)
    return UV_EBUSY;

**Comment by brimworks** [comment]
This error is triggered in this code path:

```c
int uv_pipe_accept(uv_pipe_t* server, uv_stream_t* client) {
    ...
    if (server->ipc) {
        ...
    } else {
        ...
        if (!(server->flags & UV__HANDLE_CLOSING)) {
            uv_pipe_queue_accept(loop, server, req, FALSE);
        }
    }
}

static void uv_pipe_queue_accept(uv_loop_t* loop, uv_pipe_t* handle,
    uv_pipe_accept_t* req, BOOL firstInstance) {
    ...
    if (!firstInstance) {
        ...
        if (uv_set_pipe_handle(loop, handle, req->pipeHandle, -1, 0)) {
            ...
        }
    }
}
```

I could set ```server->handle = INVALID_HANDLE_VALUE;``` before ```uv_pipe_queue_accept()``` is called and fix this, but I don't know if that is the right thing to do.

**Comment by fippo** [comment]
I just had a nice chat with @saghul about this which probably triggered a lot of memories for him :-)
Here is what I ended up with, might be useful for someone:
```
  uv_udp_t* socket = new uv_udp_t;
  uv_udp_init_ex(Loop::GetLoop(), socket, UV_UDP_RECVMMSG | AF_INET);

  uv_os_fd_t fd;
  if (uv_fileno(reinterpret_cast<uv_handle_t*>(socket), &fd) == 0) {
    uint32_t yes = 1;
    int err = setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, (char*)&yes, sizeof yes);
    printf("setsockopt SO_REUSEPORT err %d\n", err);
  }
  struct sockaddr_in recv_addr;
  uv_ip4_addr(listenIp.c_str(), listenPort, &recv_addr);
  uv_udp_bind(socket, (const struct sockaddr*)&recv_addr, UV_UDP_REUSEADDR);
  // ...

```
(for some implementations of Loop::GetLoop()

**Pull request merged**